util/website.py

```diff
@@ -0,0 +1,189 @@
+                for dir in dirs:
+                    if not sftp.isdir(dir):
+                        print "Creating directory", dir
+                        sftp.makedirs(dir)
+                        created.append(dir)
+
+                # Push all files, chdir() for each
+                for fil in files:
+                    with sftp.cd(os.path.dirname(fil)):
+                        print "Pushing", fil
+                        sftp.put(fil)
+                        pushed.append(fil)
+
+        return pushed
+    except paramiko.ssh_exception.AuthenticationException as ex:
+        print "Authentication error, please verify credentials"
+        return False
+    except IOError as ex:
+        print "Unable to read a file (private key or invalid path on server ?)", ex
+        return False
+
+    # Should not be reached
+    return False
+
+def maybe_publish(file='index.htm'):
+    """
+    Publishing to the web requires the following env variables to be set
+        'ds_website_username: defines the SSH username to connect to
+        'ds_website_privkey: defines the SSH privkey filename to use for connection
+        'ds_website_server_fqdn: hostname of the SSH server
+        'ds_website_server_port: port of the SSH server, defaults to 22
+        'ds_website_server_root: directory on the server where to push data,
+                               should be either empty, or containing at least
+                               a .htaccess and index.htm file
+    """
+
+    ssh_auth_infos = {
+        'ds_website_username': '',
+        'ds_website_privkey': '',
+        'ds_website_server_fqdn': '',
+        'ds_website_server_port': 22,
+        'ds_website_server_root': ''
+    }
+
+    for key in ssh_auth_infos.keys():
+       vartype = type(ssh_auth_infos[key])
+       value = os.environ.get(key)
+       if value is not None:
+           if vartype == str:
+               ssh_auth_infos[key] = str(os.environ.get(key))
+           elif vartype == int:
+               try:
+                   ssh_auth_infos[key] = int(os.environ.get(key))
+               except TypeError as ex:
+                   print "WARNING:", "Keeping default SSH port value because of:", ex
+
+    missing_env = filter(lambda x: len(str(ssh_auth_infos[x])) == 0, ssh_auth_infos.keys())
+    if len(missing_env) > 0:
+        print "Not publishing, missing some required environment variables:", missing_env
+        print "But maybe this is what you wanted, after all ..."
+        return False
+
+    merge_logs("logs")
+
+    all_deps = parse_for_deps(file)
+
+    if len(all_deps) == 0:
+        print "Problem during deps computation, aborting"
+        return False
+
+    if not verify_ssh_dir(ssh_auth_infos):
+        print "Problem during SSH directory verification, aborting"
+        return False
+
+    all_files = [ file ] + all_deps
+    uploaded = push_files_sftp(all_files, ssh_auth_infos)
```

Comment:
It is re-entrant. When it is already populated, verify_ssh_dir() will succeed because we do have index.htm and .htaccess files. The purpose of verify_ssh_dir is to ensure we push to a directory that makes sense, since it can be changed by env variable. So we verify if it is empty, or if it looks like an already populated website dir.

Regarding outdated dependencies, I also though about wiping but I then thought it might be leading to more problem than solutions for now: we can easily hand-remove the deprecated dependencies when we upgrade them.